Create rule S7116: The first element of an array should not be accessed implicitly CPP-5674 (#4377)
SonarSource/rspec: rules/S7116/cfamily/rule.adoc

instead of dereference (``++*array++``) or arrow (``++array->mem++``).

== Why is this an issue?

When an expression refers to a variable or another object of array type,
it usually produces a pointer to the first element of the array. 
This behavior is known as array-to-pointer decay or simply array decay.

Because of array-to-pointer decay, it is possible to use the dereference operator (`*`) or
the pointer member access operator (``++->++``) on an array object to access the first element.
However, it remains unclear if access to such elements was done intentionally.
In contrast, using array subscript (`[0]`) removes any such ambiguity.

This rule raises an issue when deference or member access operator is used
on array object.

[source,c]
----
struct Struct { int x; };
Struct elems[10];
int integers[5]; 

void init() {
  for (int i = 0; i < 5; ++i) {
    *integers = i; // Noncomliant: `*` used to access `integers[0]`
  }
}

void prcess() {
  for (int i = 0; i < 10; ++i) {
    Struct* elem = elems + i;
    elems->x += elems[i].x; // Noncomplaint: `->` used to access `elems[0]`
  }
}
----

== How to fix it

> The fix depends on the intended effect of the code, but either
> 
> * access of the first element was intentional, and uses of ``++->++`` or `*` should be replaced with subscripts with index `0`;
> * other element was meant to be accessed, and subscript appropriate index value may be used;
> * pointer values were meant to be dereferenced; in that case, the array should be replaced with the given pointer.

Reviewer: ```suggestion
The fix depends on the intended effect of the code:

* If the intent is to access the first element, ``++->++`` or `*` should be replaced with an explicit subscripts with `[0]`;
* If another element was meant to be accessed, a subscript with the appropriate index should be used;
* If the intent was to dereference a pointer (for instance to iterate on array elements), the array should be replaced with the appropriate pointer.
```

=== Code examples

==== Noncompliant code example

[source,c,diff-id=1,diff-type=noncompliant]
----
struct Struct { int x; };
Struct elems[10];

> void prcess() {

Reviewer: ```suggestion
void process() {
```

  for (int i = 0; i < 10; ++i) {

>     Struct* elem = elems + i;

Reviewer: I would only add this line if the third compliant solution
```suggestion
```

Reviewer 2: I wanted to capture the "typo" nature of the problem in this example, but now I tend to agree that this is only adding confusion. I will use that with your suggestions.

>     elems->x += elems[i].x; // Noncomplaint: -> used to access elems[0]

Reviewer: ```suggestion
    elems->x += elems[i].x; // Noncompliant: -> used to access elems[0]
```

  }
}
----

==== Compliant solution

If the access to the first element was intentional, you can use subscript explicitly:

> [source,cpp,diff-id=1,diff-type=compliant]

Reviewer: Since there are different compliant solutions, we cannot use diff for all of them, and using diff for one of them only will make it stand out too much.
```suggestion
[source,cpp]
```

Reviewer 2: I unresolved it. I want this example to stand out, as it show the change that does not change the meaning of the code, and what I intended as default quick-fix.

Reviewer: Yet in the next example, you highlight a solution that does change the meaning of the code... (I also considered removing it there, but that solution really seemed more meaningful than the other, while here all suggestions look as likely to me)

Maybe we can wait and see how it looks in the product, I'm afraid it will not look good.

Reviewer 2: Yes, I wanted the examples to hight different solutions. I used multiple solutions with diff before, and they look OK for me.

----

> struct Struct { int x; };
> Struct elems[10];

Reviewer: Maybe we can omit those declarations of the compliant examples?
```suggestion
```

Reviewer 2: This will show as removed into diff on the page, which I think will be confusing, so we cannot omit it.

Reviewer 2: And also it would be inconsitient with other rules that uses diff and repeat all declarations.

Reviewer: My opinion on this topic is not very strong.

> void prcess() {

Reviewer: ```suggestion
void process() {
```

  for (int i = 0; i < 10; ++i) {

>     Struct* elem = elems + i;

Reviewer: ```suggestion
```

>     elems[0].x += elems[i].x; // Complaint: subscript is used to access array

Reviewer: ```suggestion
    elems[0].x += elems[i].x; // Compliant: element accessed through subscript
```

  }
}
----

Otherwise, if ``++i++``th element was to be updated:
[source,cpp]
----
struct Struct { int x; };
Struct elems[10];

> void prcess() {

Reviewer: ```suggestion
void process() {
```

  for (int i = 0; i < 10; ++i) {

>     Struct* elem = elems + i;

Reviewer: ```suggestion
```

>     elems[i].x += elems[i].x; // Complaint: subscript is used to access array

Reviewer: ```suggestion
    elems[i].x += elems[i].x; // Compliant: element accessed through subscript
```

  }
}
----

Finally, the ``++i++``th element can be accessed using ``++->++`` and `elem` pointer:
[source,cpp]
----
struct Struct { int x; };
Struct elems[10];

> void prcess() {

Reviewer: ```suggestion
void process() {
```

  for (int i = 0; i < 10; ++i) {
    Struct* elem = elems + i;

>     elem->x += elem->x; // Complaint: dereferencing an pointer

Reviewer: ```suggestion
    elem->x += elem->x; // Compliant: dereferencing a pointer
```

> void prcess() {
>   for (int i = 0; i < 10; ++i) {
>     Struct* elem = elems + i;
>     elem->x += elem->x; // Complaint: dereferencing an pointer
>   }
> }

Reviewer: Should we have something more "realistic":
```suggestion
void prcess() {
  for (Struct* elem = elems;  elem < elems+10; ++elem) {
    elem->x += elem->x; // Compliant: dereferencing an pointer
  }
}
```

Reviewer 2: Changed to use something very similar.

----

==== Noncompliant code example

[source,cpp,diff-id=2,diff-type=noncompliant]
----
int integers[5]; 

void init() {
  for (int i = 0; i < 5; ++i) {

>     *integers = i; // Noncomliant: `*` used to access `integers[0]`

Reviewer: ```suggestion
    *integers = i; // Noncompliant: `*` used to access `integers[0]`
```

  }
}
----

==== Compliant solution

> Access ``++i++``th element of `integers` as intended:

Reviewer: ```suggestion
The intent was probably to access the ``++i++``th element of `integers`:
```

[source,cpp,diff-id=2,diff-type=compliant]
----
int integers[5]; 

void init() {
  for (int i = 0; i < 5; ++i) {

>     integers[i] = i; // Complaint: subscript is used to access array

Reviewer: ```suggestion
    integers[i] = i; // Compliant: element accessed through subscript
```

  }
}
----

> Or, to preserve current behavior, use subscript to access the first element:

Reviewer: ```suggestion
Or, to preserve current behavior, use a subscript to explicitly access the first element:
```

[source,cpp]
----
int integers[5]; 

void init() {
  for (int i = 0; i < 5; ++i) {

>     integers[0] = i; // Complaint: subscript is used to access array

Reviewer: ```suggestion
    integers[0] = i; // Compliant: element accessed through subscript
```

  }
}
----

== Resources

=== Documentation

* {cpp} reference - https://en.cppreference.com/w/cpp/language/array#Array-to-pointer_decay[Array-to-pointer decay]

> 

Reviewer: ```suggestion
 * S945 detects array decays that happen when passing an array object to a function that has a pointer parameter.
```
